ERROR: Failed building wheel for S15lib

Hi, when installing pyS15 package in a virtual environment (3.10.2) I get the following error:

  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for S15lib
Failed to build S15lib
ERROR: Failed to build installable wheels for some pyproject.toml based projects (S15lib)

What is wrong?

the complete listing:

(.venv) user123@AernoutvR TDC1_Project % pip3 install git+https://github.com/s-fifteen-instruments/pyS15.git                       
Collecting git+https://github.com/s-fifteen-instruments/pyS15.git
  Cloning https://github.com/s-fifteen-instruments/pyS15.git to /private/var/folders/j6/9pyrzkjs2lx_lc1vrpgr4qzh0000gn/T/pip-req-build-wg0dpjek
  Running command git clone --filter=blob:none --quiet https://github.com/s-fifteen-instruments/pyS15.git /private/var/folders/j6/9pyrzkjs2lx_lc1vrpgr4qzh0000gn/T/pip-req-build-wg0dpjek
  Resolved https://github.com/s-fifteen-instruments/pyS15.git to commit bcf4278700c483713bcd8912b9419454b63af260
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Collecting psutil (from S15lib==0.2.0)
  Using cached psutil-7.0.0-cp36-abi3-macosx_11_0_arm64.whl.metadata (22 kB)
Collecting pyserial (from S15lib==0.2.0)
  Using cached pyserial-3.5-py2.py3-none-any.whl.metadata (1.6 kB)
Collecting numpy (from S15lib==0.2.0)
  Using cached numpy-2.2.6-cp310-cp310-macosx_14_0_arm64.whl.metadata (62 kB)
Using cached numpy-2.2.6-cp310-cp310-macosx_14_0_arm64.whl (5.3 MB)
Using cached psutil-7.0.0-cp36-abi3-macosx_11_0_arm64.whl (239 kB)
Using cached pyserial-3.5-py2.py3-none-any.whl (90 kB)
Building wheels for collected packages: S15lib
  Building wheel for S15lib (pyproject.toml) ... error
  error: subprocess-exited-with-error
  
  × Building wheel for S15lib (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [730 lines of output]
      running bdist_wheel
      running build
      running build_py
      creating build/lib.macosx-10.9-universal2-cpython-310/S15lib
      copying S15lib/__init__.py -> build/lib.macosx-10.9-universal2-cpython-310/S15lib
      creating build/lib.macosx-10.9-universal2-cpython-310/S15lib/g2lib
      copying S15lib/g2lib/parse_timestamps.py -> build/lib.macosx-10.9-universal2-cpython-310/S15lib/g2lib
      copying S15lib/g2lib/__init__.py -> build/lib.macosx-10.9-universal2-cpython-310/S15lib/g2lib
      copying S15lib/g2lib/g2lib.py -> build/lib.macosx-10.9-universal2-cpython-310/S15lib/g2lib
      creating build/lib.macosx-10.9-universal2-cpython-310/S15lib/instruments
      copying S15lib/instruments/lcr_driver.py -> build/lib.macosx-10.9-universal2-cpython-310/S15lib/instruments
      copying S15lib/instruments/stepper_motor_driver.py -> build/lib.macosx-10.9-universal2-cpython-310/S15lib/instruments
      copying S15lib/instruments/powermeter.py -> build/lib.macosx-10.9-universal2-cpython-310/S15lib/instruments
      copying S15lib/instruments/spdc_driver.py -> build/lib.macosx-10.9-universal2-cpython-310/S15lib/instruments
      copying S15lib/instruments/timestamp7.py -> build/lib.macosx-10.9-universal2-cpython-310/S15lib/instruments
      copying S15lib/instruments/serial_connection.py -> build/lib.macosx-10.9-universal2-cpython-310/S15lib/instruments
      copying S15lib/instruments/digital_pattern_generator.py -> build/lib.macosx-10.9-universal2-cpython-310/S15lib/instruments
      copying S15lib/instruments/__init__.py -> build/lib.macosx-10.9-universal2-cpython-310/S15lib/instruments
      copying S15lib/instruments/usb_counter_fpga.py -> build/lib.macosx-10.9-universal2-cpython-310/S15lib/instruments
      copying S15lib/instruments/universal_discriminator.py -> build/lib.macosx-10.9-universal2-cpython-310/S15lib/instruments
      copying S15lib/instruments/single_photon_detector.py -> build/lib.macosx-10.9-universal2-cpython-310/S15lib/instruments
      creating build/lib.macosx-10.9-universal2-cpython-310/S15lib/apps
      copying S15lib/apps/__init__.py -> build/lib.macosx-10.9-universal2-cpython-310/S15lib/apps
      copying S15lib/apps/powermeter_app.py -> build/lib.macosx-10.9-universal2-cpython-310/S15lib/apps
      copying S15lib/apps/align_source.py -> build/lib.macosx-10.9-universal2-cpython-310/S15lib/apps
      copying S15lib/apps/inst_efficiency.py -> build/lib.macosx-10.9-universal2-cpython-310/S15lib/apps
      running build_ext
      building 'S15lib.g2lib.delta' extension
      creating build/temp.macosx-10.9-universal2-cpython-310/S15lib/g2lib
      /usr/bin/cc -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -arch arm64 -arch x86_64 -g -DNPY_NO_DEPRECATED_API=NPY_1_7_API_VERSION -I/private/var/folders/j6/9pyrzkjs2lx_lc1vrpgr4qzh0000gn/T/pip-build-env-sxdhk0r_/overlay/lib/python3.10/site-packages/numpy/_core/include -I/Users/user123/Python/TDC1_Project/.venv/include -I/Library/Frameworks/Python.framework/Versions/3.10/include/python3.10 -c S15lib/g2lib/delta.c -o build/temp.macosx-10.9-universal2-cpython-310/S15lib/g2lib/delta.o
      /usr/bin/cc -shared build/temp.macosx-10.9-universal2-cpython-310/S15lib/g2lib/delta.o -o build/lib.macosx-10.9-universal2-cpython-310/S15lib/g2lib/delta.cpython-310-darwin.so
      Undefined symbols for architecture arm64:
        "_PyBaseObject_Type", referenced from:
            ___Pyx_setup_reduce in delta.o
            ___pyx_memoryview_is_slice in delta.o
            ___pyx_memoryview_setitem_slice_assignment in delta.o
            ___pyx_tp_new_array in delta.o
            ___Pyx_PyErr_GivenExceptionMatches in delta.o
            ___Pyx_PyObject_FastCallDict in delta.o
            ___pyx_tp_new_Enum in delta.o
            ...
        "_PyBool_Type", referenced from:
            ___pyx_pw_6S15lib_5g2lib_5delta_9_delta_loop_ts in delta.o
        "_PyBuffer_Release", referenced from:
            ___pyx_tp_dealloc_memoryview in delta.o
        "_PyBytes_FromString", referenced from:
            ___pyx_memoryview_convert_item_to_object in delta.o
            ___pyx_memoryview_assign_item_from_object in delta.o
            ___pyx_memoryview_copy_new_contig in delta.o
        "_PyBytes_FromStringAndSize", referenced from:
            ___pyx_pymod_exec_delta in delta.o
            ___pyx_memoryview_convert_item_to_object in delta.o
        "_PyBytes_Type", referenced from:
            ___pyx_memoryview_assign_item_from_object in delta.o
            ___pyx_tp_new_array in delta.o
        "_PyCFunction_Type", referenced from:
            ___Pyx_PyObject_FastCallDict in delta.o
        "_PyCMethod_New", referenced from:
            ___pyx_pymod_exec_delta in delta.o
        "_PyCapsule_GetPointer", referenced from:
            ___pyx_pymod_exec_delta in delta.o
            ___Pyx_MergeVtables in delta.o
        "_PyCapsule_New", referenced from:
            ___pyx_pymod_exec_delta in delta.o
        "_PyCapsule_Type", referenced from:
            ___pyx_pymod_exec_delta in delta.o
        "_PyCode_NewEmpty", referenced from:
            ___Pyx_AddTraceback in delta.o
        "_PyCode_NewWithPosOnlyArgs", referenced from:
            ___pyx_pymod_exec_delta in delta.o
        "_PyDict_DelItem", referenced from:
            ___Pyx_setup_reduce in delta.o
        "_PyDict_GetItemString", referenced from:
            ___pyx_pymod_exec_delta in delta.o
        "_PyDict_New", referenced from:
            ___pyx_pymod_exec_delta in delta.o
            ___Pyx_ImportDottedModule in delta.o
            ___Pyx_CyFunction_get_dict in delta.o
            ___Pyx_CyFunction_get_annotations in delta.o
            ___pyx_pw_15View_dot_MemoryView_1__pyx_unpickle_Enum in delta.o
            ___pyx_pw_6S15lib_5g2lib_5delta_1_delta_loop in delta.o
            ___pyx_pw_6S15lib_5g2lib_5delta_3_cond_delta_loop in delta.o
            ...
        "_PyDict_Next", referenced from:
            ___Pyx_CyFunction_CallAsMethod in delta.o
            ___pyx_pw___pyx_array_1__reduce_cython__ in delta.o
            ___Pyx_ParseOptionalKeywords in delta.o
            ___pyx_pw___pyx_MemviewEnum_1__reduce_cython__ in delta.o
            ___pyx_memoryview_is_c_contig in delta.o
            ___pyx_memoryview_is_f_contig in delta.o
            ___pyx_memoryview_copy in delta.o
            ...
        "_PyDict_SetItem", referenced from:
            ___pyx_pymod_exec_delta in delta.o
            ___Pyx_setup_reduce in delta.o
            ___pyx_pw_6S15lib_5g2lib_5delta_1_delta_loop in delta.o
            ___pyx_pw_6S15lib_5g2lib_5delta_3_cond_delta_loop in delta.o
            ___pyx_pw_6S15lib_5g2lib_5delta_9_delta_loop_ts in delta.o
        "_PyDict_SetItemString", referenced from:
            ___pyx_pymod_create in delta.o
            ___pyx_pymod_exec_delta in delta.o
        "_PyDict_Size", referenced from:
            ___Pyx_CyFunction_CallMethod in delta.o
            ___pyx_tp_new_array in delta.o
            ___pyx_MemviewEnum___init__ in delta.o
            ___pyx_tp_new_memoryview in delta.o
        "_PyErr_Clear", referenced from:
            ___pyx_pymod_create in delta.o
            ___pyx_pymod_exec_delta in delta.o
            ___Pyx_AddTraceback in delta.o
            ___Pyx_ImportDottedModule in delta.o
            ___Pyx_CyFunction_get_is_coroutine in delta.o
            ___Pyx_MergeVtables in delta.o
            ___Pyx_setup_reduce in delta.o
            ...
        "_PyErr_ExceptionMatches", referenced from:
            ___pyx_pymod_create in delta.o
            ___pyx_pymod_exec_delta in delta.o
            ___pyx_tp_getattro_array in delta.o
            ___Pyx_GetItemInt_Fast in delta.o
            ___Pyx_SetItemInt_Fast in delta.o
            ___pyx_pw_15View_dot_MemoryView_1__pyx_unpickle_Enum in delta.o
        "_PyErr_Fetch", referenced from:
            ___pyx_tp_dealloc_array in delta.o
            ___pyx_tp_dealloc_memoryview in delta.o
            ___pyx_tp_dealloc__memoryviewslice in delta.o
        "_PyErr_Format", referenced from:
            ___pyx_pymod_exec_delta in delta.o
            ___Pyx_CyFunction_CallAsMethod in delta.o
            ___Pyx_CyFunction_CallMethod in delta.o
            ___Pyx_PyType_Ready in delta.o
            ___Pyx_MergeVtables in delta.o
            ___Pyx_setup_reduce in delta.o
            ___pyx_memoryview_setitem_slice_assignment in delta.o
            ...
        "_PyErr_GivenExceptionMatches", referenced from:
            ___Pyx_PyObject_GetItem in delta.o
            ___Pyx_PyErr_GivenExceptionMatches in delta.o
        "_PyErr_NoMemory", referenced from:
            ___pyx_pymod_exec_delta in delta.o
            ___Pyx_CyFunction_CallAsMethod in delta.o
            ___pyx_memoryview_setitem_slice_assign_scalar in delta.o
            ___pyx_memoryview_copy_contents in delta.o
        "_PyErr_NormalizeException", referenced from:
            ___Pyx__GetException in delta.o
        "_PyErr_Occurred", referenced from:
            ___pyx_pymod_exec_delta in delta.o
            ___Pyx_ImportDottedModule in delta.o
            ___pyx_array_get_memview in delta.o
            ___Pyx_MergeVtables in delta.o
            ___Pyx_setup_reduce in delta.o
            ___pyx_memoryview_get_item_pointer in delta.o
            ___pyx_memoryview_is_slice in delta.o
            ...
        "_PyErr_Restore", referenced from:
            ___pyx_tp_dealloc_array in delta.o
            ___pyx_tp_dealloc_memoryview in delta.o
            ___pyx_tp_dealloc__memoryviewslice in delta.o
        "_PyErr_SetObject", referenced from:
            ___Pyx_Raise in delta.o
        "_PyErr_SetString", referenced from:
            ___pyx_pymod_create in delta.o
            ___pyx_pymod_exec_delta in delta.o
            ___Pyx_CyFunction_New in delta.o
            ___Pyx_CyFunction_CallAsMethod in delta.o
            ___Pyx_CyFunction_CallMethod in delta.o
            ___Pyx_CyFunction_set_name in delta.o
            ___Pyx_CyFunction_set_qualname in delta.o
            ...
        "_PyErr_WarnEx", referenced from:
            ___pyx_pymod_exec_delta in delta.o
            ___Pyx_CyFunction_set_defaults in delta.o
            ___Pyx_CyFunction_set_kwdefaults in delta.o
            ___Pyx_ImportType_3_0_12 in delta.o
        "_PyErr_WarnFormat", referenced from:
            ___Pyx_PyNumber_IntOrLong in delta.o
        "_PyExc_AttributeError", referenced from:
            ___pyx_pymod_create in delta.o
            ___pyx_pymod_exec_delta in delta.o
            ___pyx_tp_getattro_array in delta.o
            ___Pyx_PyObject_GetAttrStrNoError in delta.o
            ___pyx_pw___pyx_MemviewEnum_1__reduce_cython__ in delta.o
            ___pyx_pw_15View_dot_MemoryView_1__pyx_unpickle_Enum in delta.o
        "_PyExc_BufferError", referenced from:
            ___pyx_array_getbuffer in delta.o
            ___pyx_memoryview_getbuffer in delta.o
        "_PyExc_DeprecationWarning", referenced from:
            ___Pyx_PyNumber_IntOrLong in delta.o
        "_PyExc_ImportError", referenced from:
            ___pyx_pymod_create in delta.o
            ___pyx_pymod_exec_delta in delta.o
            ___pyx_pw_15View_dot_MemoryView_1__pyx_unpickle_Enum in delta.o
        "_PyExc_IndexError", referenced from:
            ___Pyx_PyObject_GetItem in delta.o
            ___pyx_memoryview_slice_memviewslice in delta.o
            ___pyx_pw_6S15lib_5g2lib_5delta_3_cond_delta_loop in delta.o
        "_PyExc_ModuleNotFoundError", referenced from:
            ___pyx_pymod_exec_delta in delta.o
        "_PyExc_NameError", referenced from:
            ___pyx_pymod_exec_delta in delta.o
            ___pyx_pw___pyx_MemviewEnum_1__reduce_cython__ in delta.o
            ___pyx_pw_6S15lib_5g2lib_5delta_1_delta_loop in delta.o
            ___pyx_pw_6S15lib_5g2lib_5delta_3_cond_delta_loop in delta.o
            ___pyx_pw_6S15lib_5g2lib_5delta_5delta_loop in delta.o
            ___pyx_pw_6S15lib_5g2lib_5delta_7cond_delta_loop in delta.o
            ___pyx_pw_6S15lib_5g2lib_5delta_9_delta_loop_ts in delta.o
            ...
        "_PyExc_NotImplementedError", referenced from:
            ___pyx_mp_ass_subscript_array in delta.o
            ___pyx_mp_ass_subscript_memoryview in delta.o
        "_PyExc_OverflowError", referenced from:
            ___pyx_memoryview_get_item_pointer in delta.o
            ___pyx_memoryview_convert_item_to_object in delta.o
            ___pyx_tp_new_array in delta.o
            ___Pyx_PyObject_GetItem in delta.o
            ___Pyx_GetItemInt_Fast in delta.o
            ___Pyx_PyUnicode_Join in delta.o
            ___Pyx_PyInt_As_int in delta.o
            ...
        "_PyExc_RuntimeError", referenced from:
            ___pyx_pymod_exec_delta in delta.o
            ___Pyx_MergeVtables in delta.o
            ___Pyx_setup_reduce in delta.o
        "_PyExc_RuntimeWarning", referenced from:
            ___Pyx_CyFunction_set_defaults in delta.o
            ___Pyx_CyFunction_set_kwdefaults in delta.o
        "_PyExc_StopIteration", referenced from:
            ___pyx_memoryview_get_item_pointer in delta.o
            ___pyx_memoryview___getitem__ in delta.o
            ___pyx_pw_6S15lib_5g2lib_5delta_11delta_loop_ts in delta.o
        "_PyExc_SystemError", referenced from:
            ___pyx_pymod_exec_delta in delta.o
            ___Pyx_CyFunction_New in delta.o
            ___Pyx_CyFunction_CallMethod in delta.o
            ___pyx_array_get_memview in delta.o
            ___pyx_memoryview_is_slice in delta.o
            ___pyx_memoryview_setitem_slice_assignment in delta.o
            ___pyx_memoryview_assign_item_from_object in delta.o
            ...
        "_PyExc_TypeError", referenced from:
            ___pyx_pymod_exec_delta in delta.o
            ___Pyx_CyFunction_CallAsMethod in delta.o
            ___Pyx_CyFunction_CallMethod in delta.o
            ___Pyx_CyFunction_set_name in delta.o
            ___Pyx_CyFunction_set_qualname in delta.o
            ___Pyx_CyFunction_set_dict in delta.o
            ___Pyx_CyFunction_set_defaults in delta.o
            ...
        "_PyExc_UnboundLocalError", referenced from:
            ___pyx_memoryview___getitem__ in delta.o
        "_PyExc_ValueError", referenced from:
            ___pyx_memoryview_copy_contents in delta.o
            ___pyx_memslice_transpose in delta.o
            ___pyx_memoryview___getitem__ in delta.o
            ___pyx_mp_ass_subscript_memoryview in delta.o
            ___pyx_memoryview_slice_memviewslice in delta.o
            ___pyx_memoryview_copy_new_contig in delta.o
            ___Pyx_init_memviewslice in delta.o
            ...
        "_PyExc_ZeroDivisionError", referenced from:
            ___pyx_memoryview_get_item_pointer in delta.o
            ___pyx_tp_new_array in delta.o
            ___pyx_pw_6S15lib_5g2lib_5delta_1_delta_loop in delta.o
            ___pyx_pw_6S15lib_5g2lib_5delta_3_cond_delta_loop in delta.o
            ___pyx_pw_6S15lib_5g2lib_5delta_9_delta_loop_ts in delta.o
        "_PyException_SetTraceback", referenced from:
            ___Pyx__GetException in delta.o
        "_PyFloat_AsDouble", referenced from:
            ___pyx_pw_6S15lib_5g2lib_5delta_1_delta_loop in delta.o
            ___pyx_pw_6S15lib_5g2lib_5delta_3_cond_delta_loop in delta.o
            ___pyx_pw_6S15lib_5g2lib_5delta_5delta_loop in delta.o
            ___pyx_pw_6S15lib_5g2lib_5delta_7cond_delta_loop in delta.o
            ___pyx_pw_6S15lib_5g2lib_5delta_9_delta_loop_ts in delta.o
            ___pyx_pw_6S15lib_5g2lib_5delta_11delta_loop_ts in delta.o
        "_PyFloat_FromDouble", referenced from:
            ___pyx_pf_6S15lib_5g2lib_5delta_12__defaults__ in delta.o
            ___pyx_pf_6S15lib_5g2lib_5delta_14__defaults__ in delta.o
            ___pyx_pf_6S15lib_5g2lib_5delta_16__defaults__ in delta.o
            ___Pyx_PyInt_AddObjC in delta.o
            ___pyx_pw_6S15lib_5g2lib_5delta_5delta_loop in delta.o
            ___pyx_pw_6S15lib_5g2lib_5delta_7cond_delta_loop in delta.o
            ___pyx_pw_6S15lib_5g2lib_5delta_11delta_loop_ts in delta.o
            ...
        "_PyFloat_Type", referenced from:
            ___pyx_tp_new_array in delta.o
            __unellipsify in delta.o
            ___pyx_pw_6S15lib_5g2lib_5delta_1_delta_loop in delta.o
            ___Pyx_PyInt_AddObjC in delta.o
            ___pyx_pw_6S15lib_5g2lib_5delta_3_cond_delta_loop in delta.o
            ___pyx_pw_6S15lib_5g2lib_5delta_5delta_loop in delta.o
            ___pyx_pw_6S15lib_5g2lib_5delta_7cond_delta_loop in delta.o
            ...
        "_PyFrame_New", referenced from:
            ___Pyx_AddTraceback in delta.o
        "_PyGC_Disable", referenced from:
            ___Pyx_PyType_Ready in delta.o
        "_PyGC_Enable", referenced from:
            ___Pyx_PyType_Ready in delta.o
        "_PyGILState_Ensure", referenced from:
            ___pyx_memoryview_setitem_slice_assign_scalar in delta.o
            ___pyx_memoryview_copy_contents in delta.o
            ___pyx_memoryview_err_dim in delta.o
            ___pyx_memslice_transpose in delta.o
            ___pyx_memoryview_slice_memviewslice in delta.o
        "_PyGILState_Release", referenced from:
            ___pyx_memoryview_setitem_slice_assign_scalar in delta.o
            ___pyx_memoryview_copy_contents in delta.o
            ___pyx_memoryview_err_dim in delta.o
            ___pyx_memslice_transpose in delta.o
            ___pyx_memoryview_slice_memviewslice in delta.o
        "_PyImport_AddModule", referenced from:
            ___pyx_pymod_exec_delta in delta.o
        "_PyImport_GetModule", referenced from:
            ___Pyx_ImportDottedModule in delta.o
            ___pyx_pw_15View_dot_MemoryView_1__pyx_unpickle_Enum in delta.o
        "_PyImport_GetModuleDict", referenced from:
            ___pyx_pymod_exec_delta in delta.o
        "_PyImport_ImportModule", referenced from:
            ___pyx_pymod_exec_delta in delta.o
        "_PyImport_ImportModuleLevelObject", referenced from:
            ___Pyx_ImportDottedModule in delta.o
            ___Pyx_CyFunction_get_is_coroutine in delta.o
            ___pyx_pw_15View_dot_MemoryView_1__pyx_unpickle_Enum in delta.o
        "_PyIndex_Check", referenced from:
            ___pyx_memoryview___getitem__ in delta.o
            __unellipsify in delta.o
        "_PyInterpreterState_GetID", referenced from:
            ___pyx_pymod_create in delta.o
        "_PyList_Append", referenced from:
            ___pyx_getprop___pyx_memoryview_shape in delta.o
            ___pyx_getprop___pyx_memoryview_strides in delta.o
            ___pyx_getprop___pyx_memoryview_suboffsets in delta.o
        "_PyList_AsTuple", referenced from:
            __unellipsify in delta.o
            ___pyx_getprop___pyx_memoryview_shape in delta.o
            ___pyx_getprop___pyx_memoryview_strides in delta.o
            ___pyx_getprop___pyx_memoryview_suboffsets in delta.o
        "_PyList_New", referenced from:
            ___Pyx_CyFunction_get_is_coroutine in delta.o
            __unellipsify in delta.o
            ___pyx_getprop___pyx_memoryview_shape in delta.o
            ___pyx_getprop___pyx_memoryview_strides in delta.o
            ___pyx_getprop___pyx_memoryview_suboffsets in delta.o
            ___pyx_pw_15View_dot_MemoryView_1__pyx_unpickle_Enum in delta.o
        "_PyList_Type", referenced from:
            ___pyx_memoryview_get_item_pointer in delta.o
            ___Pyx_GetItemInt_Fast in delta.o
            ___pyx_memoryview___getitem__ in delta.o
            ___Pyx_SetItemInt_Fast in delta.o
            ___pyx_pw_6S15lib_5g2lib_5delta_11delta_loop_ts in delta.o
        "_PyLong_AsLong", referenced from:
            ___Pyx_PyInt_As_int in delta.o
            ___Pyx_PyInt_As_long in delta.o
        "_PyLong_AsSsize_t", referenced from:
            ___pyx_memoryview_get_item_pointer in delta.o
            ___pyx_tp_new_array in delta.o
            ___Pyx_PyObject_GetItem in delta.o
            ___pyx_memoryview___getitem__ in delta.o
            ___pyx_memoryview_fromslice in delta.o
        "_PyLong_FromDouble", referenced from:
            ___pyx_pw_6S15lib_5g2lib_5delta_1_delta_loop in delta.o
            ___pyx_pw_6S15lib_5g2lib_5delta_3_cond_delta_loop in delta.o
            ___pyx_pw_6S15lib_5g2lib_5delta_9_delta_loop_ts in delta.o
        "_PyLong_FromLong", referenced from:
            ___pyx_pymod_exec_delta in delta.o
            ___pyx_array_get_memview in delta.o
            ___pyx_memoryview_is_slice in delta.o
            ___pyx_memoryview_err_dim in delta.o
            ___pyx_memoryview_new in delta.o
            ___pyx_getprop___pyx_memoryview_ndim in delta.o
            ___pyx_pw_15View_dot_MemoryView_1__pyx_unpickle_Enum in delta.o
            ...
        "_PyLong_FromSsize_t", referenced from:
            ___pyx_sq_item_array in delta.o
            ___Pyx_GetItemInt_Fast in delta.o
            ___pyx_unpickle_Enum__set_state in delta.o
            ___pyx_sq_item_memoryview in delta.o
            __unellipsify in delta.o
            ___Pyx_SetItemInt_Fast in delta.o
            ___pyx_memoryview_fromslice in delta.o
            ...
        "_PyLong_Type", referenced from:
            ___pyx_pymod_exec_delta in delta.o
            ___pyx_memoryview_get_item_pointer in delta.o
            ___pyx_tp_new_array in delta.o
            ___Pyx_PyObject_GetItem in delta.o
            ___Pyx_PyNumber_IntOrLong in delta.o
            ___pyx_memoryview___getitem__ in delta.o
            __unellipsify in delta.o
            ...
        "_PyMem_Free", referenced from:
            ___Pyx_CyFunction_CallAsMethod in delta.o
            ___pyx_memoryview_setitem_slice_assign_scalar in delta.o
        "_PyMem_Malloc", referenced from:
            ___Pyx_AddTraceback in delta.o
            ___Pyx_CyFunction_CallAsMethod in delta.o
            ___pyx_memoryview_setitem_slice_assign_scalar in delta.o
        "_PyMem_Realloc", referenced from:
            ___Pyx_AddTraceback in delta.o
        "_PyMethod_New", referenced from:
            ___Pyx_PyMethod_New in delta.o
        "_PyMethod_Type", referenced from:
            ___pyx_memoryview_convert_item_to_object in delta.o
            ___pyx_memoryview_assign_item_from_object in delta.o
            ___pyx_tp_new_array in delta.o
            ___pyx_unpickle_Enum__set_state in delta.o
            ___pyx_pw_15View_dot_MemoryView_1__pyx_unpickle_Enum in delta.o
            ___pyx_pw_6S15lib_5g2lib_5delta_5delta_loop in delta.o
            ___pyx_pw_6S15lib_5g2lib_5delta_7cond_delta_loop in delta.o
            ...
        "_PyModuleDef_Init", referenced from:
            _PyInit_delta in delta.o
        "_PyModule_GetDict", referenced from:
            ___pyx_pymod_create in delta.o
            ___pyx_pymod_exec_delta in delta.o
        "_PyModule_GetName", referenced from:
            ___pyx_pw_15View_dot_MemoryView_1__pyx_unpickle_Enum in delta.o
        "_PyModule_NewObject", referenced from:
            ___pyx_pymod_create in delta.o
        "_PyNumber_Add", referenced from:
            ___pyx_memoryview_assign_item_from_object in delta.o
        "_PyNumber_InPlaceAdd", referenced from:
            ___pyx_pw___pyx_MemviewEnum_1__reduce_cython__ in delta.o
            ___Pyx_PyInt_AddObjC in delta.o
        "_PyNumber_InPlaceMultiply", referenced from:
            ___pyx_memoryview_fromslice in delta.o
            ___pyx_getprop___pyx_memoryview_size in delta.o
        "_PyNumber_Index", referenced from:
            ___pyx_memoryview_get_item_pointer in delta.o
            ___pyx_tp_new_array in delta.o
            ___Pyx_PyObject_GetItem in delta.o
            ___pyx_memoryview___getitem__ in delta.o
            ___pyx_memoryview_fromslice in delta.o
        "_PyNumber_Multiply", referenced from:
            ___pyx_getprop___pyx_memoryview_suboffsets in delta.o
            ___pyx_getprop___pyx_memoryview_nbytes in delta.o
        "_PyNumber_Remainder", referenced from:
            ___pyx_memoryview_err_dim in delta.o
        "_PyOS_snprintf", referenced from:
            ___pyx_pymod_exec_delta in delta.o
            ___Pyx_ImportType_3_0_12 in delta.o
        "_PyObject_Call", referenced from:
            ___pyx_pymod_exec_delta in delta.o
            ___pyx_array_get_memview in delta.o
            ___pyx_memoryview_is_slice in delta.o
            ___pyx_memoryview_assign_item_from_object in delta.o
            ___Pyx_Raise in delta.o
            ___Pyx_PyObject_FastCallDict in delta.o
            ___pyx_memoryview_new in delta.o
            ...
        "_PyObject_CallFinalizerFromDealloc", referenced from:
            ___pyx_tp_dealloc_array in delta.o
            ___pyx_tp_dealloc_Enum in delta.o
            ___pyx_tp_dealloc_memoryview in delta.o
            ___pyx_tp_dealloc__memoryviewslice in delta.o
        "_PyObject_ClearWeakRefs", referenced from:
            ___Pyx_CyFunction_dealloc in delta.o
        "_PyObject_Format", referenced from:
            ___pyx_tp_new_array in delta.o
            __unellipsify in delta.o
        "_PyObject_Free", referenced from:
            ___Pyx_CyFunction_clear in delta.o
            ___pyx_tp_dealloc_array in delta.o
        "_PyObject_GC_Del", referenced from:
            ___Pyx_CyFunction_dealloc in delta.o
        "_PyObject_GC_IsFinalized", referenced from:
            ___pyx_tp_dealloc_array in delta.o
            ___pyx_tp_dealloc_Enum in delta.o
            ___pyx_tp_dealloc_memoryview in delta.o
            ___pyx_tp_dealloc__memoryviewslice in delta.o
        "_PyObject_GC_Track", referenced from:
            ___Pyx_CyFunction_New in delta.o
            ___pyx_tp_dealloc__memoryviewslice in delta.o
        "_PyObject_GC_UnTrack", referenced from:
            ___Pyx_CyFunction_dealloc in delta.o
            ___pyx_tp_dealloc_Enum in delta.o
            ___pyx_tp_dealloc_memoryview in delta.o
            ___pyx_tp_dealloc__memoryviewslice in delta.o
        "_PyObject_GenericGetAttr", referenced from:
            ___pyx_pymod_exec_delta in delta.o
            ___pyx_tp_getattro_array in delta.o
            ___Pyx_PyObject_GetAttrStrNoError in delta.o
        "_PyObject_GetAttr", referenced from:
            ___pyx_pymod_exec_delta in delta.o
            ___Pyx_CyFunction_get_is_coroutine in delta.o
            ___Pyx_setup_reduce in delta.o
            ___pyx_memoryview_setitem_slice_assignment in delta.o
            ___pyx_memoryview_convert_item_to_object in delta.o
            ___pyx_memoryview_assign_item_from_object in delta.o
            ___pyx_tp_new_array in delta.o
            ...
        "_PyObject_GetAttrString", referenced from:
            ___pyx_pymod_create in delta.o
            ___pyx_pymod_exec_delta in delta.o
            ___Pyx_ImportType_3_0_12 in delta.o
        "_PyObject_GetBuffer", referenced from:
            ___pyx_tp_new_memoryview in delta.o
        "_PyObject_GetItem", referenced from:
            ___Pyx_MergeVtables in delta.o
            ___Pyx_GetItemInt_Fast in delta.o
            ___pyx_unpickle_Enum__set_state in delta.o
        "_PyObject_GetIter", referenced from:
            ___pyx_memoryview_get_item_pointer in delta.o
            ___pyx_memoryview___getitem__ in delta.o
            ___pyx_pw_6S15lib_5g2lib_5delta_11delta_loop_ts in delta.o
        "_PyObject_Hash", referenced from:
            ___pyx_pymod_exec_delta in delta.o
        "_PyObject_IsInstance", referenced from:
            ___pyx_pw_6S15lib_5g2lib_5delta_5delta_loop in delta.o
        "_PyObject_IsSubclass", referenced from:
            ___Pyx_Raise in delta.o
        "_PyObject_IsTrue", referenced from:
            ___pyx_pymod_exec_delta in delta.o
            ___Pyx_ImportDottedModule in delta.o
            ___pyx_tp_new_array in delta.o
            ___Pyx_PyUnicode_Equals in delta.o
            ___pyx_tp_new_memoryview in delta.o
            ___pyx_memoryview___getitem__ in delta.o
            ___pyx_mp_ass_subscript_memoryview in delta.o
            ...
        "_PyObject_Malloc", referenced from:
            ___pyx_pymod_exec_delta in delta.o
            ___pyx_tp_new_array in delta.o
        "_PyObject_Not", referenced from:
            ___Pyx_AddTraceback in delta.o
        "_PyObject_RichCompare", referenced from:
            ___pyx_pymod_exec_delta in delta.o
            ___pyx_tp_new_array in delta.o
            ___Pyx_PyUnicode_Equals in delta.o
        "_PyObject_RichCompareBool", referenced from:
            ___Pyx_setup_reduce in delta.o
        "_PyObject_SetAttr", referenced from:
            ___pyx_pymod_exec_delta in delta.o
            ___Pyx_AddTraceback in delta.o
        "_PyObject_SetAttrString", referenced from:
            ___pyx_pymod_exec_delta in delta.o
        "_PyObject_SetItem", referenced from:
            ___pyx_mp_ass_subscript_array in delta.o
            __unellipsify in delta.o
            ___Pyx_SetItemInt_Fast in delta.o
            ___pyx_pw_6S15lib_5g2lib_5delta_1_delta_loop in delta.o
            ___pyx_pw_6S15lib_5g2lib_5delta_3_cond_delta_loop in delta.o
            ___pyx_pw_6S15lib_5g2lib_5delta_9_delta_loop_ts in delta.o
        "_PyObject_Size", referenced from:
            ___pyx_pw_6S15lib_5g2lib_5delta_5delta_loop in delta.o
            ___pyx_pw_6S15lib_5g2lib_5delta_7cond_delta_loop in delta.o
            ___pyx_pw_6S15lib_5g2lib_5delta_11delta_loop_ts in delta.o
        "_PyObject_VectorcallDict", referenced from:
            ___Pyx_PyObject_FastCallDict in delta.o
        "_PySequence_Contains", referenced from:
            ___pyx_pw_15View_dot_MemoryView_1__pyx_unpickle_Enum in delta.o
        "_PySequence_Tuple", referenced from:
            ___pyx_memoryview_assign_item_from_object in delta.o
        "_PySlice_New", referenced from:
            ___pyx_pymod_exec_delta in delta.o
        "_PySlice_Type", referenced from:
            __unellipsify in delta.o
        "_PyThreadState_Get", referenced from:
            ___pyx_pymod_create in delta.o
        "_PyThread_allocate_lock", referenced from:
            ___pyx_pymod_exec_delta in delta.o
        "_PyThread_free_lock", referenced from:
            ___pyx_tp_dealloc_memoryview in delta.o
        "_PyTraceBack_Here", referenced from:
            ___Pyx_AddTraceback in delta.o
        "_PyTuple_GetItem", referenced from:
            ___Pyx_CyFunction_CallAsMethod in delta.o
        "_PyTuple_GetSlice", referenced from:
            ___Pyx_CyFunction_CallAsMethod in delta.o
        "_PyTuple_New", referenced from:
            ___pyx_pymod_exec_delta in delta.o
            ___pyx_pf_6S15lib_5g2lib_5delta_12__defaults__ in delta.o
            ___pyx_pf_6S15lib_5g2lib_5delta_14__defaults__ in delta.o
            ___pyx_pf_6S15lib_5g2lib_5delta_16__defaults__ in delta.o
            ___Pyx_CyFunction_CallAsMethod in delta.o
            ___pyx_array_get_memview in delta.o
            ___pyx_memoryview_get_item_pointer in delta.o
            ...
        "_PyTuple_Pack", referenced from:
            ___pyx_pymod_exec_delta in delta.o
            ___Pyx_Raise in delta.o
        "_PyTuple_Type", referenced from:
            ___pyx_memoryview_get_item_pointer in delta.o
            ___pyx_memoryview_assign_item_from_object in delta.o
            ___pyx_tp_new_array in delta.o
            ___Pyx_GetItemInt_Fast in delta.o
            ___pyx_pw___pyx_MemviewEnum_3__setstate_cython__ in delta.o
            ___pyx_memoryview___getitem__ in delta.o
            ___pyx_pw_15View_dot_MemoryView_1__pyx_unpickle_Enum in delta.o
            ...
        "_PyType_Modified", referenced from:
            ___pyx_pymod_exec_delta in delta.o
            ___Pyx_setup_reduce in delta.o
        "_PyType_Ready", referenced from:
            ___pyx_pymod_exec_delta in delta.o
            ___Pyx_PyType_Ready in delta.o
        "_PyUnicode_AsUTF8", referenced from:
            ___Pyx_AddTraceback in delta.o
        "_PyUnicode_Compare", referenced from:
            ___Pyx_ParseOptionalKeywords in delta.o
        "_PyUnicode_Concat", referenced from:
            ___pyx_tp_new_array in delta.o
            ___pyx_pw_15View_dot_MemoryView_1__pyx_unpickle_Enum in delta.o
        "_PyUnicode_Decode", referenced from:
            ___pyx_pymod_exec_delta in delta.o
        "_PyUnicode_Format", referenced from:
            ___pyx_memoryview_err_dim in delta.o
            ___pyx_memoryview___repr__ in delta.o
            ___pyx_memoryview___str__ in delta.o
            ___pyx_pw_15View_dot_MemoryView_1__pyx_unpickle_Enum in delta.o
        "_PyUnicode_FromFormat", referenced from:
            ___Pyx_AddTraceback in delta.o
            ___Pyx_CyFunction_repr in delta.o
        "_PyUnicode_FromOrdinal", referenced from:
            ___pyx_memoryview_get_item_pointer in delta.o
            ___pyx_tp_new_array in delta.o
            ___pyx_memoryview_copy_contents in delta.o
        "_PyUnicode_FromString", referenced from:
            ___Pyx_CyFunction_get_doc in delta.o
            ___pyx_pw_15View_dot_MemoryView_1__pyx_unpickle_Enum in delta.o
        "_PyUnicode_FromStringAndSize", referenced from:
            ___pyx_pymod_exec_delta in delta.o
        "_PyUnicode_InternFromString", referenced from:
            ___pyx_pymod_exec_delta in delta.o
            ___Pyx_CyFunction_get_name in delta.o
        "_PyUnicode_New", referenced from:
            ___pyx_memoryview_get_item_pointer in delta.o
            ___pyx_tp_new_array in delta.o
            ___Pyx_PyUnicode_Join in delta.o
            ___pyx_memoryview_copy_contents in delta.o
        "_PyUnicode_Type", referenced from:
            ___pyx_tp_new_array in delta.o
            ___Pyx_PyUnicode_Equals in delta.o
            ___pyx_memoryview_err_dim in delta.o
            __unellipsify in delta.o
        "_Py_EnterRecursiveCall", referenced from:
            ___pyx_pymod_exec_delta in delta.o
            ___pyx_array_get_memview in delta.o
            ___pyx_memoryview_is_slice in delta.o
            ___pyx_memoryview_assign_item_from_object in delta.o
            ___Pyx_PyObject_FastCallDict in delta.o
            ___pyx_memoryview_new in delta.o
            ___pyx_pw_6S15lib_5g2lib_5delta_1_delta_loop in delta.o
            ...
        "_Py_GetVersion", referenced from:
            ___pyx_pymod_exec_delta in delta.o
        "_Py_LeaveRecursiveCall", referenced from:
            ___pyx_pymod_exec_delta in delta.o
            ___pyx_array_get_memview in delta.o
            ___pyx_memoryview_is_slice in delta.o
            ___pyx_memoryview_assign_item_from_object in delta.o
            ___Pyx_PyObject_FastCallDict in delta.o
            ___pyx_memoryview_new in delta.o
            ___pyx_pw_6S15lib_5g2lib_5delta_1_delta_loop in delta.o
            ...
        "_Py_OptimizeFlag", referenced from:
            ___pyx_tp_new_memoryview in delta.o
            ___pyx_memoryview___getitem__ in delta.o
        "__PyDict_GetItem_KnownHash", referenced from:
            ___pyx_pymod_exec_delta in delta.o
            ___Pyx_AddTraceback in delta.o
            ___pyx_tp_new_array in delta.o
            ___pyx_MemviewEnum___init__ in delta.o
            ___pyx_pw___pyx_MemviewEnum_1__reduce_cython__ in delta.o
            ___pyx_tp_new_memoryview in delta.o
            ___pyx_pw_6S15lib_5g2lib_5delta_1_delta_loop in delta.o
            ...
        "__PyObject_GC_New", referenced from:
            ___Pyx_CyFunction_New in delta.o
        "__PyObject_GenericGetAttrWithDict", referenced from:
            ___Pyx_PyObject_GetAttrStrNoError in delta.o
        "__PyObject_GetDictPtr", referenced from:
            ___Pyx_AddTraceback in delta.o
        "__PyThreadState_UncheckedGet", referenced from:
            ___pyx_pymod_exec_delta in delta.o
            ___Pyx_AddTraceback in delta.o
            ___pyx_memoryview_is_slice in delta.o
            ___pyx_memoryview_setitem_slice_assign_scalar in delta.o
            ___pyx_memoryview_convert_item_to_object in delta.o
            ___Pyx_PyObject_GetAttrStrNoError in delta.o
            ___pyx_pw___pyx_MemviewEnum_1__reduce_cython__ in delta.o
            ...
        "__PyType_Lookup", referenced from:
            ___Pyx_setup_reduce in delta.o
        "__PyUnicode_FastCopyCharacters", referenced from:
            ___Pyx_PyUnicode_Join in delta.o
        "__PyUnicode_Ready", referenced from:
            ___Pyx_PyUnicode_Equals in delta.o
            ___Pyx_PyUnicode_Join in delta.o
        "__Py_Dealloc", referenced from:
            ___pyx_pymod_create in delta.o
            ___pyx_pymod_exec_delta in delta.o
            ___Pyx_AddTraceback in delta.o
            ___Pyx__GetException in delta.o
            ___Pyx_ImportDottedModule in delta.o
            ___Pyx_CyFunction_New in delta.o
            ___pyx_pf_6S15lib_5g2lib_5delta_12__defaults__ in delta.o
            ...
        "__Py_FalseStruct", referenced from:
            ___pyx_pymod_exec_delta in delta.o
            ___Pyx_AddTraceback in delta.o
            ___Pyx_ImportDottedModule in delta.o
            ___Pyx_CyFunction_get_is_coroutine in delta.o
            ___pyx_array_get_memview in delta.o
            ___pyx_memoryview_is_slice in delta.o
            ___pyx_tp_new_array in delta.o
            ...
        "__Py_FatalErrorFunc", referenced from:
            ___pyx_fatalerror in delta.o
        "__Py_NoneStruct", referenced from:
            ___pyx_pymod_create in delta.o
            ___pyx_pymod_exec_delta in delta.o
            ___Pyx_ImportDottedModule in delta.o
            ___pyx_pf_6S15lib_5g2lib_5delta_12__defaults__ in delta.o
            ___pyx_pf_6S15lib_5g2lib_5delta_14__defaults__ in delta.o
            ___pyx_pf_6S15lib_5g2lib_5delta_16__defaults__ in delta.o
            ___Pyx_CyFunction_get_doc in delta.o
            ...
        "__Py_TrueStruct", referenced from:
            ___pyx_pymod_exec_delta in delta.o
            ___Pyx_AddTraceback in delta.o
            ___Pyx_ImportDottedModule in delta.o
            ___Pyx_CyFunction_get_is_coroutine in delta.o
            ___pyx_array_get_memview in delta.o
            ___pyx_memoryview_is_slice in delta.o
            ___pyx_tp_new_array in delta.o
            ...
      ld: symbol(s) not found for architecture arm64
      clang: error: linker command failed with exit code 1 (use -v to see invocation)
      error: command '/usr/bin/cc' failed with exit code 1
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for S15lib
Failed to build S15lib
ERROR: Failed to build installable wheels for some pyproject.toml based projects (S15lib)

--- a/README.md
+++ b/README.md
@@ -8,7 +8,11 @@ S15lib is a Python package for controlling [S-Fifteen instruments](https://s-fif
 Install the package directly with
 
 ```
+# For Linux/MacOS
 pip install git+https://github.com/s-fifteen-instruments/pyS15.git
+
+# For Windows and/or systems with compile-related installation difficulties
+pip install git+https://github.com/s-fifteen-instruments/pyS15.git@no_compile
 ```
 
 Alternatively, clone or [download](https://github.com/s-fifteen-instruments/pyS15/archive/refs/heads/master.zip)

--- a/setup.py
+++ b/setup.py
@@ -27,30 +27,23 @@ requirements_apps = [
 #
 # Precedence:
 #   1. User environment variable, e.g. CC=gcc
-#   2. Python build-time compiler
-#   3. OS executable in PATH, e.g. /usr/bin/cc
+#   2. OS executable in PATH, e.g. /usr/bin/cc
+#   3. Python build-time compiler
 #
 # TODO: Check if compatible with Windows installations
-config = sysconfig.get_config_vars()
-
-# Check if Python was cross-compiled (i.e. build-time compiler does not exist)
-def tokenize(command):  # noqa
-    return shlex.split(command)
 
 
 def get_executable_index(tokens, check_exists=True):
     for i, token in enumerate(tokens):
-        # Ignore environment variables, and check executable exists
-        if "=" not in token:
+        if "=" not in token:  # ignore environment variables
             if check_exists and not shutil.which(token):
                 continue
             return i
-    else:
-        return None
+    return None
 
 
 def replace_executable(target, command):
-    tokens = tokenize(command)
+    tokens = shlex.split(command)
     idx = get_executable_index(tokens, check_exists=False)
     tokens = list(tokens)  # replace with mutable container
     tokens[idx] = target
@@ -58,53 +51,80 @@ def replace_executable(target, command):
 
 
 def find_executable(command):
-    tokens = tokenize(command)
+    tokens = shlex.split(command)
     idx = get_executable_index(tokens)
     return idx is not None
 
 
-u_cc = os.environ.get("CC", None)  # user-specified compiler
-if u_cc is not None:
-    config["CC"] = u_cc
-    config["LDSHARED"] = f"{u_cc} -shared"  # for shared libraries
-    u_ldshared = os.environ.get("LDSHARED", None)
-    if u_ldshared is not None:
-        config["LDSHARED"] = u_ldshared  # if manual library linking required
+def get_compiler(config=sysconfig.get_config_vars(), env={}):
+    """Returns preferred cc and ldshared programs.
 
-else:
-    b_cc = config.get("CC", None)  # build compiler
-    b_ldshared = config.get("LDSHARED", None)
-    p_cc = shutil.which("cc")  # platform default compiler (w/o options)
-    if b_cc is None:
-        config["CC"] = p_cc
-        config["LDSHARED"] = f"{p_cc} -shared"
-    elif not find_executable(b_cc):
-        config["CC"] = replace_executable(p_cc, b_cc)
-        config["LDSHARED"] = replace_executable(p_cc, b_ldshared)
+    Args:
+        config: config dictionary from Python's sysconfig.
+        env: environment variables.
 
+    Examples:
+        $ export PROG="python -c 'import setup, os; \
+            print(setup.get_compiler(setup.config, os.environ))'"
+
+        # Assuming gcc build compiler, not present on host
+        $ eval $PROG
+        ('cc -pthread', 'cc -pthread -shared -Wl,--exclude-libs,ALL -LModules/_hacl')
+
+        # Creating a symlink named 'gcc'
+        $ ln -s /usr/bin/cc /usr/bin/gcc
+        $ eval $PROG
+        ('gcc -pthread', 'gcc -pthread -shared -Wl,--exclude-libs,ALL -LModules/_hacl')
+
+        # Use env-defined compilers
+        $ CC='clang -Wall' eval $PROG
+        ('clang -Wall', 'clang -Wall -shared')
+        $ CC=clang LDSHARED='clang -shared -O2' eval $PROG
+        ('clang', 'clang -shared -O2')
+    """
+
+    # Prefer user-supplied compiler (flags)
+    if "CC" in env:
+        cc = env.get("CC")
+        return cc, env.get("LDSHARED", f"{cc} -shared")
+
+    platform_cc = shutil.which("cc")
+    cc = config.get("CC", platform_cc)
+    ldshared = config.get("LDSHARED", f"{platform_cc} -shared")
+    if not find_executable(cc):  # replace build compiler with platform's
+        return (
+            replace_executable(platform_cc, cc),
+            replace_executable(platform_cc, ldshared),
+        )
+    return cc, ldshared  # use default compiler
+
+
+config = sysconfig.get_config_vars()
+config["CC"], config["LDSHARED"] = get_compiler(config, os.environ)
 
 # setuptools project configuration
-setuptools.setup(
-    name="S15lib",
-    version="0.2.0",
-    description="S-Fifteen Python Library",
-    url="https://s-fifteen.com/",
-    author="Mathias Seidler;",
-    author_email="",
-    license="MIT",
-    packages=setuptools.find_packages(),
-    install_requires=requirements,
-    extras_require={
-        "dev": requirements_dev,
-        "apps": requirements_apps,
-    },
-    python_requires=">=3.6",
-    ext_modules=[
-        setuptools.Extension(
-            name="S15lib.g2lib.delta",
-            sources=["S15lib/g2lib/delta.c"],
-            include_dirs=[np.get_include()],
-            define_macros=[("NPY_NO_DEPRECATED_API", "NPY_1_7_API_VERSION")],
-        )
-    ],
-)
+if __name__ == "__main__":
+    setuptools.setup(
+        name="S15lib",
+        version="0.2.0",
+        description="S-Fifteen Python Library",
+        url="https://s-fifteen.com/",
+        author="Mathias Seidler;",
+        author_email="",
+        license="MIT",
+        packages=setuptools.find_packages(),
+        install_requires=requirements,
+        extras_require={
+            "dev": requirements_dev,
+            "apps": requirements_apps,
+        },
+        python_requires=">=3.6",
+        ext_modules=[
+            setuptools.Extension(
+                name="S15lib.g2lib.delta",
+                sources=["S15lib/g2lib/delta.c"],
+                include_dirs=[np.get_include()],
+                define_macros=[("NPY_NO_DEPRECATED_API", "NPY_1_7_API_VERSION")],
+            )
+        ],
+    )

--- a/setup.py
+++ b/setup.py
@@ -1,4 +1,5 @@
 import os
+import shlex
 import shutil
 import sysconfig
 
@@ -26,16 +27,60 @@ requirements_apps = [
 #
 # Precedence:
 #   1. User environment variable, e.g. CC=gcc
-#   2. OS executable in PATH, e.g. /usr/bin/cc
-#   3. Python build-time compiler
+#   2. Python build-time compiler
+#   3. OS executable in PATH, e.g. /usr/bin/cc
+#
+# TODO: Check if compatible with Windows installations
 config = sysconfig.get_config_vars()
-cc = os.environ.get("CC", shutil.which("cc"))
-if cc is not None:
-    config["CC"] = cc
-    config["LDSHARED"] = f"{cc} -shared"  # for shared libraries
-ldshared = os.environ.get("LDSHARED", None)  # allow user override
-if ldshared is not None:
-    config["LDSHARED"] = ldshared
+
+# Check if Python was cross-compiled (i.e. build-time compiler does not exist)
+def tokenize(command):  # noqa
+    return shlex.split(command)
+
+
+def get_executable_index(tokens, check_exists=True):
+    for i, token in enumerate(tokens):
+        # Ignore environment variables, and check executable exists
+        if "=" not in token:
+            if check_exists and not shutil.which(token):
+                continue
+            return i
+    else:
+        return None
+
+
+def replace_executable(target, command):
+    tokens = tokenize(command)
+    idx = get_executable_index(tokens, check_exists=False)
+    tokens = list(tokens)  # replace with mutable container
+    tokens[idx] = target
+    return shlex.join(tokens)
+
+
+def find_executable(command):
+    tokens = tokenize(command)
+    idx = get_executable_index(tokens)
+    return idx is not None
+
+
+u_cc = os.environ.get("CC", None)  # user-specified compiler
+if u_cc is not None:
+    config["CC"] = u_cc
+    config["LDSHARED"] = f"{u_cc} -shared"  # for shared libraries
+    u_ldshared = os.environ.get("LDSHARED", None)
+    if u_ldshared is not None:
+        config["LDSHARED"] = u_ldshared  # if manual library linking required
+
+else:
+    b_cc = config.get("CC", None)  # build compiler
+    b_ldshared = config.get("LDSHARED", None)
+    p_cc = shutil.which("cc")  # platform default compiler (w/o options)
+    if b_cc is None:
+        config["CC"] = p_cc
+        config["LDSHARED"] = f"{p_cc} -shared"
+    elif not find_executable(b_cc):
+        config["CC"] = replace_executable(p_cc, b_cc)
+        config["LDSHARED"] = replace_executable(p_cc, b_ldshared)
 
 
 # setuptools project configuration

